contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,498 @@
+            emit Executed(id);
+        }
+    }
+
+    function _cancel(bytes32 id) internal virtual {
+        require(_schedules[id].isSet(), "invalid schedule");
+        _schedules[id] = 0.toTimepoint(); // delete
+        emit Canceled(id);
+    }
+
+    function _hashOperation(address caller, address target, bytes calldata data) internal virtual pure returns (bytes32) {
+        return keccak256(abi.encode(caller, target, data));
+    }
+}
+
+
+
+
+contract AccessManager is IAccessManager, DelayedActions {
+    using Time for *;
+
+    bytes32 public constant ADMIN_GROUP  = bytes32(type(uint256).min); // 0
+    bytes32 public constant PUBLIC_GROUP = bytes32(type(uint256).max); // 2**256-1
+
+    mapping(address target => AccessMode mode) private _contractMode;
+    mapping(address target => mapping(bytes4 selector => bytes32 group)) private _allowedGroups;
+    mapping(bytes32 group => Group) private _groups;
+
+    /**
+     * @dev Check that the caller has a given permission level (`group`). Note that this does NOT consider execution
+     * delays that may be associated to that group.
+     */
+    modifier onlyGroup(bytes32 group) {
+        require(hasGroup(group, _msgSender()));
+        _;
+    }
+
+    constructor(address initialAdmin) {
+        // admin is active immediatly and without any execution delay.
+        _grantRole(ADMIN_GROUP, initialAdmin, 0.toDuration(), 0.toDuration());
+    }
+
+    // =================================================== GETTERS ====================================================
+    /**
+     * @dev Check if an address (`caller`) is authorised to call a given function on a given contract. Additionally,
+     * returns the delay needed to perform the call. If there is a delay, the call must be scheduled (with {schedule})
+     * and executed (with {relay}).
+     *
+     * This function is usually called by the targeted contract to control immediate execution of restricted functions.
+     * Therefore we only return true is the call can be performed without any delay. If the call is subject to a delay,
+     * then the function should return false, and the caller should schedule the operation for future execution.
+     *
+     * We may be able to hash the operation, and check if the call was scheduled, but we would not be able to cleanup
+     * the schedule, leaving the possibility of multiple executions. Maybe this function should not be view?
+     */
+    function canCall(address caller, address target, bytes4 selector) public view virtual returns (bool, uint32) {
+        AccessMode mode = getContractMode(target);
+        if (mode == AccessMode.Open) {
+            return (true, 0);
+        } else if (mode == AccessMode.Closed) {
+            return (false, Time.MAX_DURATION.get()); // use 0 ?
+        } else if (caller == address(this)) {
+            // Caller is AccessManager => call was relayed. In that case the relay already checked permissions.
+            return (true, 0);
```

Comment:
This is concerning. I think we should have a flag in storage (transient storage, eventually) during a relay, and only return true if the flag is set, possibly we should only return true for the specific target/selector combination that was relayed.